aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -1309,4 +1309,20 @@ module aptos_framework::account {
         account_state.guid_creation_num = MAX_GUID_CREATION_NUM - 1;
         create_guid(account);
     }
+
+    #[test_only]
+    struct FakeCoin { }
+
+    #[test(account = @0x1234)]
+    fun test_events(account: &signer) acquires Account {
+        let addr = signer::address_of(account);
+        create_account_unchecked(addr);
+        register_coin<FakeCoin>(addr);
+
+        let eventhandle = &borrow_global<Account>(addr).coin_register_events;
+        let counter = event::counter(eventhandle);
```

Comment:
The need for the counter appears unfortunate to me. You would never ever have to remember it apart of for this case. Also if the event is emitted somewhere deep inside some call chain, how to even get the counter?

Why not return a vector of all events? Then you can write something like:

```
assert!(contains(event:emitted_events(handle), |e| ...e...))
```

Checking for the counter becomes a special case.

Reply:
> The need for the counter appears unfortunate to me. You would never ever have to remember it apart of for this case. Also if the event is emitted somewhere deep inside some call chain, how to even get the counter?
> 
> Why not return a vector of all events? Then you can write something like:
> 
> ```
> assert!(contains(event:emitted_events(handle), |e| ...e...))
> ```
> 
> Checking for the counter becomes a special case.

Another problem: both with your and mine suggestion event needs to be `copy` which usually it isn't. Damn it we do not have `Option<&T>` or `vector<&T>`. If we would have fulll lambdas this could be mitigated: `event:emitted(|e: &E| ...)`

Reply:
These are intended for unit tests. So having explicit knowledge seems relevant. That's also why I didn't expose the native function, that way we can always change it to be more expressive as you suggest, add more accessors for all events, and get an easy bump.

Regardless of where we land here, I think we'll want to have the access by sequence number.

Reply:
I don't understand why this explicit knowledge of the sequence number would be relevant. For events v2 were are in the process of removing those, so why would anybody care?

Also, how can you know the sequence number if the event is emitted somewhere deep inside the logic you are testing? Lets say your test function calls A calls B calls C which emits the event. How would you know the number? The code you are testing doesn't remember the number as it doesn't care about it. How to get it up to the test?

The issue have having events emitted somewhere quite indirectly deep inside code is well-known from the time we implemented event verification for the Move prover.

Last not least, testing based on sequence number is not compostional. Suppose I test a function which calls A and then B. B emits an event and I test it. A today does not emit an event. But now I change the code so A also does emit an even. Now your test fails because the sequence number shifted. Not good.

Reply:
I've added a new core function that returns all (seq_num, events)